### PR TITLE
Implementation of decision tree classifier using pytorch backend

### PR DESCRIPTION
**Pull request by dikshant2210**
- [x] Docstrings.
- [x] Working DecisionTreeClassifier
- [x] Tests
- [ ] Notebook.

### CODE REVIEW

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            if size == 0:
+                continue
+            score = 0.0
+            for cl in classes:
+                proportion = sum(y == cl) / size
+                score += proportion ** 2
+            gini += (1.0 - score) * (size / num_instances)
+        return gini
+
+    # TODO: Implement entropy loss
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: remove

File: omega/DecisionTree/DecisionTreeRegressor.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeRegressor.py]: remove

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        mask = torch.nonzero(1 - torch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            right = (torch.index_select(x, 0, mask), torch.index_select(y, 0, mask))
+        else:
+            right = (torch.DoubleTensor(), torch.DoubleTensor())
+        return left, right
+
+    def fit(self, x, y):
+        if isinstance(x, np.ndarray):
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Write tests for this.

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            x = torch.from_numpy(x)
+        if isinstance(y, np.ndarray):
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Write tests for this.

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+# Copyright (C) 2017 Dikshant Gupta <dikshant2210@gmail.com>
+# Licensed under the GNU General Public License v3.0 - https://www.gnu.org/licenses/gpl-3.0.en.html
+
+import torch
+import numpy as np
+
+
+class DecisionTreeClassifier:
+
+    def __init__(self,
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: use hanging indentsl

File: omega/test/test_dt.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                nodes.append(nodes[0]['left'])
+                nodes.append(nodes[0]['right'])
+
+            nodes = nodes[1:]
+            if not nodes:
+                break
+        self.model_equal(observed, expected)
+
+    def test_min_size(self):
+        dt = DecisionTreeClassifier(2, 6)
```
Comment (markroxor) [inline omega/test/test_dt.py]: Explain what does this tests do with comments.

File: omega/test/test_dt.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        y1 = torch.Tensor([0])
+
+        x2 = torch.Tensor([10.12493903, 3.234550982])
+        y2 = torch.Tensor([1])
+
+        self.assertEqual(y1[0], dt.predict(x1))
+        self.assertEqual(y2[0], dt.predict(x2))
+
+    def test_numpy_input(self):
+        dt = DecisionTreeClassifier(2, 1)
```
Comment (markroxor) [inline omega/test/test_dt.py]: Explain what does this tests do with comments.

File: omega/test/test_dt.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                nodes.append(nodes[0]['left'])
+                nodes.append(nodes[0]['right'])
+
+            nodes = nodes[1:]
+            if not nodes:
+                break
+        self.model_equal(observed, expected)
+
+    def test_predict(self):
+        dt = DecisionTreeClassifier(2, 1)
```
Comment (markroxor) [inline omega/test/test_dt.py]: Explain what does this tests do with comments.

File: omega/test/test_dt.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                nodes.append(nodes[0]['left'])
+                nodes.append(nodes[0]['right'])
+
+            nodes = nodes[1:]
+            if not nodes:
+                break
+        self.model_equal(observed, expected)
+
+    def test_model_depth_3(self):
+        dt = DecisionTreeClassifier(3, 1)
```
Comment (markroxor) [inline omega/test/test_dt.py]: Explain what does this tests do with comments.

File: omega/test/test_dt.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                               [7.444542326, 0.476683375],
+                               [10.12493903, 3.234550982],
+                               [6.642287351, 3.319983761]])
+        self.y = torch.Tensor([0, 0, 0, 0, 0, 1, 1, 1, 1, 1])
+
+    def model_equal(self, observed, expected):
+        self.assertEqual(observed, expected)
+
+    def test_model_depth_2(self):
+        dt = DecisionTreeClassifier(2, 1)
```
Comment (markroxor) [inline omega/test/test_dt.py]: Explain what does this tests do with comments.

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        else:
+            node['right'] = self.__get_split(right[0], right[1])
+            self.__split(node['right'], depth + 1)
+
+    def predict(self, row):
+        """Given a row as data, predicts the class of the dataset.
+
+        Parameters
+        ----------
+        row : row similar to the training data
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+        """
+        res = torch.mode(group[1])[0][0]
+        return res
+
+    def __split(self, node, depth):
+        """Build the subtree given a node, keeps track of maximum depth of the tree.
+
+        Parameters
+        ----------
+        node : node for which the subtree is built
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+        for index in range(x.size()[1]):
+            for row in x:
+                groups = self.__test_split(index, row[index], x, y)
+                score = self.__gini_index(groups[0], groups[1], classes)
+                if score < b_score:
+                    b_index, b_value, b_score, b_groups = index, row[index], score, groups
+        return {'index': b_index, 'value': b_value, 'groups': b_groups}
+
+    def __to_terminal(self, group):
+        """Determines the label of a terminal node.
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+        Returns
+        -------
+        A dictionary containing the tree.
+        """
+        root = self.__get_split(x, y)
+        self.__split(root, 1)
+        return root
+
+    def __get_split(self, x, y):
+        """Finds the best split given the training data and labels
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+        if isinstance(y, np.ndarray):
+            y = torch.from_numpy(y)
+        if self.__use_cuda:
+            x = x.cuda()
+            y = y.cuda()
+        self.__root = self.__build_tree(x, y)
+        return self.__root
+
+    def __build_tree(self, x, y):
+        """Initiates the tree building process by calling for the first split.
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        mask = torch.nonzero(1 - torch.lt(x[:, index], value))
+        if mask.size():
+            mask = mask.view(mask.size()[0])
+            right = (torch.index_select(x, 0, mask), torch.index_select(y, 0, mask))
+        else:
+            right = (torch.DoubleTensor(), torch.DoubleTensor())
+        return left, right
+
+    def fit(self, x, y):
+        """Trains the classifier on the training data and forms the decision tree
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+                continue
+            score = 0.0
+            for cl in classes:
+                proportion = sum(y == cl) / size
+                score += proportion ** 2
+            gini += (1.0 - score) * (size / num_instances)
+        return gini
+
+    def __test_split(self, index, value, x, y):
+        """Splits the dataset into two parts on the basis of a value.
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+            Represents the metrics to compare each split point.
+        """
+        self.__use_cuda = torch.cuda.is_available()
+        self.__max_depth = max_depth
+        self.__min_size = min_size
+        self.__split_metric = split_metric
+        self.__root = {}
+
+    def __gini_index(self, left, right, classes):
+        """Calculates the value of gini index for each split.
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Sphinx will not process this docstring, check the docstrings in my code and follow
https://pythonhosted.org/an_example_pypi_project/buildanduploadsphinx.html

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    def __init__(self,
+                 max_depth=100,
+                 min_size=2,
+                 split_metric='gini_index'):
+        """Initializes the parameters of a decision tree and create root node
+        the decision as an empty dictionary.
+
+        Parameters
+        ----------
+        max_depth : int, optional, default value 100
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: This is the correct way of writing doc strings.

**Comment by markroxor** [comment]
Merge master in your branch and check if the docs are built as expected.

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+    >>> import torch
+    >>> x = torch.Tensor([[2.771244718, 1.784783929], [1.728571309, 1.169761413], [7.444542326, 0.476683375])
+    >>> y = torch.Tensor([0, 0 ,1])
+    >>> dt = DecisionTreeClassifier()
+    >>> dt.fit(x, y)
+    >>> dt.predict(torch.Tensor([10.12493903, 3.234550982]))
+    1.0
+    """
+
+    def __init__(self,
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: Use hanging indents.
https://www.python.org/dev/peps/pep-0008/#indentation

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+                score += proportion ** 2
+            gini += (1.0 - score) * (size / num_instances)
+        return gini
+
+    def __test_split(self, index, value, x, y):
+        """Splits the dataset into two parts on the basis of a value.
+
+        Parameters
+        ----------
+        index :  int, index of the attribute on which to split
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: fix this

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+
+        Parameters
+        ----------
+        left :  ch.Tensor
+            a fraction of the original data
+        right : ch.Tensor
+            a fraction of the original data
+        classes : list
+            list of all possible values to target labels.
+        """
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: returns???

File: omega/DecisionTree/DecisionTreeClassifier.py [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+            training data
+        y : ch.tensor
+            training labels
+
+        Returns
+        -------
+        dict : dict
+            A node for the tree with best split value given the data.
+        """
+        b_index, b_value, b_score, b_groups = 999, 999, 999, None
```
Comment (markroxor) [inline omega/DecisionTree/DecisionTreeClassifier.py]: What is this?

Comment (dikshant2210) [reply omega/DecisionTree/DecisionTreeClassifier.py]: random initializations, i just need the variables initialized outside the loop

**Pull request merged**